homeassistant/components/hifiberry/media_player.py

```diff
@@ -0,0 +1,262 @@
+
+HifiBerry rest API: https://github.com/hifiberry/audiocontrol2/blob/fee165140c9da044c2166cac1d77ae5e0008c351/doc/api.md
+"""
+
+from datetime import timedelta
+import logging
+import socket
+import requests
+
+import voluptuous as vol
+
+from homeassistant.components.media_player import PLATFORM_SCHEMA, MediaPlayerDevice
+from homeassistant.components.media_player.const import (
+    MEDIA_TYPE_MUSIC,
+    SUPPORT_NEXT_TRACK,
+    SUPPORT_PAUSE,
+    SUPPORT_PLAY,
+    SUPPORT_PREVIOUS_TRACK,
+    SUPPORT_STOP,
+    SUPPORT_VOLUME_MUTE,
+    SUPPORT_VOLUME_SET,
+    SUPPORT_VOLUME_STEP,
+)
+from homeassistant.const import (
+    CONF_HOST,
+    CONF_NAME,
+    CONF_PORT,
+    STATE_IDLE,
+    STATE_PAUSED,
+    STATE_PLAYING,
+)
+import homeassistant.helpers.config_validation as cv
+
+_LOGGER = logging.getLogger(__name__)
+
+DEFAULT_NAME = "HifiBerry"
+DEFAULT_PORT = 81
+
+DATA_HIFIBERRY = "hifiberry"
+
+TIMEOUT = 10
+SCAN_INTERVAL = timedelta(seconds=1)
+
+SUPPORT_HIFIBERRY = (
+    SUPPORT_PAUSE
+    | SUPPORT_VOLUME_SET
+    | SUPPORT_VOLUME_MUTE
+    | SUPPORT_PREVIOUS_TRACK
+    | SUPPORT_NEXT_TRACK
+    | SUPPORT_STOP
+    | SUPPORT_PLAY
+    | SUPPORT_VOLUME_STEP
+)
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
+    {
+        vol.Required(CONF_HOST): cv.string,
+        vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
+        vol.Optional(CONF_PORT, default=DEFAULT_PORT): cv.port,
+    }
+)
+
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """Set up the HifiBerry platform."""
+    if DATA_HIFIBERRY not in hass.data:
+        hass.data[DATA_HIFIBERRY] = dict()
+
+    # This is a manual configuration?
+    if discovery_info is None:
+        name = config.get(CONF_NAME)
+        host = config.get(CONF_HOST)
+        port = config.get(CONF_PORT)
+    else:
+        name = "{} ({})".format(DEFAULT_NAME, discovery_info.get("hostname"))
+        host = discovery_info.get("host")
+        port = discovery_info.get("port")
+
+    # Only add a device once, so discovered devices do not override manual
+    # config.
+    ip_addr = socket.gethostbyname(host)
+    if ip_addr in hass.data[DATA_HIFIBERRY]:
+        return
+
+    device = HifiBerry(name, host, port, hass)
+
+    hass.data[DATA_HIFIBERRY][ip_addr] = device
+    add_devices([device])
+
+
+class HifiBerry(MediaPlayerDevice):
+    """HifiBerry Player Object."""
+
+    def __init__(self, name, host, port, hass):
+        """Initialize the media player."""
+        self.host = host
+        self.port = port
+        self.hass = hass
+        self._url = "{}:{}".format(host, str(port))
+        self._name = name
+        self._muted = False
+        self._muted_volume = 0
+        self._state = {}
+
+    def get_hifiberry_msg(self, method, params=None):
+        """Send message."""
+        url = f"http://{self.host}:{self.port}/{method}"
+        _LOGGER.debug("URL: %s params: %s", url, params)
+        response = requests.get(url, params=params)
```

Comment:
Please extract the device specific interface code and data parsing to a standalone library published on PyPI.

https://developers.home-assistant.io/docs/creating_component_code_review#4-communication-with-devicesservices